cumulusci/tasks/apex/testrunner.py

```diff
@@ -314,16 +360,61 @@ def _run_task(self):
             self.classes_by_name[test_class["Name"]] = test_class["Id"]
             self.results_by_class_name[test_class["Name"]] = {}
         self.logger.info("Queuing tests for execution...")
-        ids = list(self.classes_by_id.keys())
-        result = self.tooling._call_salesforce(
-            method="POST",
-            url=self.tooling.base_url + "runTestsAsynchronous",
-            json={"classids": ",".join(str(id) for id in ids)},
+
+        self.counts = {
+            "Pass": 0,
+            "Fail": 0,
+            "CompileFail": 0,
+            "Skip": 0,
+            "Retriable": 0,
+        }
+        self.job_id = self._enqueue_test_run(
+            (str(id) for id in self.classes_by_id.keys())
         )
-        self.job_id = result.json()
+
         self._wait_for_tests()
-        test_results = self._get_test_results()
+        self._get_test_results()
+
+        # Did we get back retriable test results? Check our retry policy,
+        # then enqueue new runs individually, until either (a) all retriable
+        # tests succeed or (b) a test fails.
+        able_to_retry = (self.counts["Retriable"] and self.options["retry_always"]) or (
+            self.counts["Retriable"] and self.counts["Retriable"] == self.counts["Fail"]
+        )
+        if not able_to_retry:
+            self.counts["Retriable"] = 0
+        else:
+            self.logger.warning(
```

Comment:
What do you think about extracting this else clause into a method like `_attempt_retries()`?

Reply:
Sold!